python/pyspark/mllib/linalg.py

```diff
@@ -1152,9 +1156,385 @@ def sparse(numRows, numCols, colPtrs, rowIndices, values):
+        if not isinstance(first, Vector):
+            raise TypeError("rows should be an RDD of Vectors, got an RDD of %s" % type(first))
+
+        javaRowMatrix = callMLlibFunc("createRowMatrix", rows, long(numRows), int(numCols))
+        self._jrm = JavaModelWrapper(javaRowMatrix)
+        self.rows = rows
+
+    @staticmethod
+    def _from_java(javaRowMatrix):
+        """Create a PySpark RowMatrix from a Java RowMatrix object."""
+        rows = JavaModelWrapper(javaRowMatrix).call("rows")
+        return RowMatrix(rows)
+
+    def numRows(self):
+        """
+        Get or compute the number of rows.
+
+        >>> rows = sc.parallelize([Vectors.dense([1, 2, 3]), Vectors.dense([4, 5, 6]),
+        ...                        Vectors.dense([7, 8, 9]), Vectors.dense([10, 11, 12])])
+        >>> rm = RowMatrix(rows)
+        >>> int(rm.numRows())
+        4
+
+        >>> rows = sc.parallelize([Vectors.dense([1, 2, 3]), Vectors.dense([4, 5, 6]),
+        ...                        Vectors.dense([7, 8, 9]), Vectors.dense([10, 11, 12])])
+        >>> rm = RowMatrix(rows, 7, 6)
+        >>> int(rm.numRows())
+        7
+        """
+        return self._jrm.call("numRows")
+
+    def numCols(self):
+        """
+        Get or compute the number of cols.
+
+        >>> rows = sc.parallelize([Vectors.dense([1, 2, 3]), Vectors.dense([4, 5, 6]),
+        ...                        Vectors.dense([7, 8, 9]), Vectors.dense([10, 11, 12])])
+        >>> rm = RowMatrix(rows)
+        >>> int(rm.numCols())
+        3
+
+        >>> rows = sc.parallelize([Vectors.dense([1, 2, 3]), Vectors.dense([4, 5, 6]),
+        ...                        Vectors.dense([7, 8, 9]), Vectors.dense([10, 11, 12])])
+        >>> rm = RowMatrix(rows, 7, 6)
+        >>> int(rm.numCols())
+        6
+        """
+        return self._jrm.call("numCols")
+
+
+class IndexedRow(object):
+    """
+    Represents a row of an IndexedRowMatrix.
+
+    Just a wrapper over a (long, Vector) tuple.
+
+    .. note:: Experimental
+    """
+    def __init__(self, index, vector):
+        self.index = long(index)
+        self.vector = _convert_to_vector(vector)
+
+    def __repr__(self):
+        return "IndexedRow(%s, %s)" % (self.index, self.vector)
+
+
+class IndexedRowMatrix(DistributedMatrix):
+    """
+    Represents a row-oriented distributed Matrix with indexed rows.
+
+    .. note:: Experimental
+    """
+    def __init__(self, rows, numRows=0, numCols=0):
+        """
+        Create a wrapper over a Java IndexedRowMatrix.
+
+        :param rows: An RDD of IndexedRows or (long, Vector) tuples.
+        """
+        if not isinstance(rows, RDD):
+            raise TypeError("rows should be an RDD object, got %s" % type(rows))
+        first = rows.first()
+        if not (isinstance(first, IndexedRow) or
+                (isinstance(first, tuple) and len(first) == 2 and
+                 isinstance(first[0], (long, int)) and isinstance(first[1], Vector))):
```

Comment:
Yeah, I'm not opposed to this, but we should still check to make sure it is 2-element `tuple` so that a useful error message is returned in the event that a completely wrong data structure is passed in.